File: src/llm/gemini-client.ts
import { GoogleGenerativeAI, SchemaType } from '@google/generative-ai';
import { A11yIssue } from '../state/types';
import { getSystemPrompt } from '../prompts';

interface AnalysisResult {
  issues: A11yIssue[];
  summary: string;
}

export class GeminiClient {
  private client: GoogleGenerativeAI;
  private model: string;

  constructor(apiKey: string, model: string = 'gemini-2.0-flash') {
    this.client = new GoogleGenerativeAI(apiKey);
    this.model = model;
  }

  async analyze(diffContent: string, prompt: string): Promise<AnalysisResult> {
    const schema = {
      type: SchemaType.OBJECT,
      properties: {
        issues: {
          type: SchemaType.ARRAY,
          items: {
            type: SchemaType.OBJECT,
            properties: {
              file: { type: SchemaType.STRING },
              line: { type: SchemaType.INTEGER, nullable: true },
              wcag_criterion: { type: SchemaType.STRING },
              wcag_level: { type: SchemaType.STRING },
              severity: { type: SchemaType.STRING },
              title: { type: SchemaType.STRING },
              description: { type: SchemaType.STRING },
              suggestion: { type: SchemaType.STRING },
            },
            required: ['file', 'wcag_criterion', 'wcag_level', 'severity', 'description', 'suggestion'],
          },
        },
        summary: { type: SchemaType.STRING },
      },
      required: ['issues', 'summary'],
    };

    const systemInstruction = getSystemPrompt();
    const fullUserPrompt = `${prompt}\n\n---\n\n## Code Diff:\n\n${diffContent}`;

    const genModel = this.client.getGenerativeModel({
      model: this.model,
      systemInstruction,
      generationConfig: {
        temperature: 0.1,
        responseMimeType: 'application/json',
        responseSchema: schema,
      },
    });

    try {
      const result = await genModel.generateContent(fullUserPrompt);
      const text = result.response.text();
      
      const parsed = JSON.parse(text);
      
      const issues: A11yIssue[] = (parsed.issues || []).map((issue: Record<string, unknown>) => {
        const rawSeverity = String(issue.severity || 'GOOD_PRACTICE').toUpperCase();
        const severity: A11yIssue['severity'] = rawSeverity === 'VIOLATION' ? 'VIOLATION' : 'GOOD_PRACTICE';
        
        return {
          file: String(issue.file || ''),
          line: issue.line ? Number(issue.line) : null,
          wcag_criterion: String(issue.wcag_criterion || ''),
          wcag_level: String(issue.wcag_level || 'A'),
          severity,
          title: String(issue.title || ''),
          description: String(issue.description || ''),
          suggestion: String(issue.suggestion || ''),
        };
      });

      return {
        issues,
        summary: String(parsed.summary || 'Accessibility review completed.'),
      };
    } catch (error) {
      throw new Error(`Gemini API error: ${error instanceof Error ? error.message : String(error)}`);
    }
  }
}

File: src/llm/gemini_client.py
"""Gemini API client for accessibility review."""

import json
import os
from typing import Dict, List, Optional, Any

from .base import LLMClient, LLMResponse


class GeminiClient(LLMClient):
    """Google Gemini API client implementation."""

    def __init__(
        self,
        api_key: Optional[str] = None,
        model: str = 'gemini-2.0-flash'
    ):
        """
        Initialize the Gemini client.

        Args:
            api_key: Google AI API key (can also be set via GEMINI_API_KEY env)
            model: Gemini model to use
        """
        self._api_key = api_key or os.getenv('GEMINI_API_KEY')
        if not self._api_key:
            raise ValueError("Gemini API key required. Set GEMINI_API_KEY environment variable or pass api_key parameter.")

        self._model_name = model
        self._client = None
        self._initialized = False

    def _ensure_client(self):
        """Lazy initialization of the Gemini client."""
        if self._initialized:
            return

        try:
            import google.generativeai as genai
        except ImportError:
            raise ImportError(
                "google-generativeai package not installed. "
                "Install it with: pip install google-generativeai"
            )

        genai.configure(api_key=self._api_key)
        self._client = genai.GenerativeModel(self._model_name)
        self._initialized = True

    @property
    def model_name(self) -> str:
        return self._model_name

    @property
    def backend_type(self) -> str:
        return 'gemini'

    def analyze_diff(
        self,
        diff_content: str,
        system_prompt: str,
        user_prompt: str,
        json_schema: Optional[Dict] = None
    ) -> LLMResponse:
        """Analyze diff using Gemini API."""
        self._ensure_client()

        import google.generativeai as genai

        # Build the full prompt
        full_prompt = f"{system_prompt}\n\n{user_prompt}\n\n## Diff to Analyze:\n\n{diff_content}"

        # Configure for JSON output if schema provided
        generation_config = {
            'temperature': 0.1,  # Low temperature for consistent analysis
            'top_p': 0.95,
        }

        if json_schema:
            # Use response_schema for structured output (Gemini 1.5+)
            generation_config['response_mime_type'] = 'application/json'
            generation_config['response_schema'] = json_schema

        try:
            model = genai.GenerativeModel(
                self._model_name,
                generation_config=generation_config
            )

            response = model.generate_content(full_prompt)

            # Extract content
            content = response.text

            # Parse structured data if JSON
            structured_data = None
            if json_schema and content:
                try:
                    parsed = json.loads(content)
                    if isinstance(parsed, list):
                        structured_data = parsed
                    elif isinstance(parsed, dict) and 'issues' in parsed:
                        structured_data = parsed['issues']
                except json.JSONDecodeError:
                    pass

            # Extract usage info
            usage = None
            if hasattr(response, 'usage_metadata'):
                usage = {
                    'prompt_tokens': getattr(response.usage_metadata, 'prompt_token_count', 0),
                    'completion_tokens': getattr(response.usage_metadata, 'candidates_token_count', 0),
                    'total_tokens': getattr(response.usage_metadata, 'total_token_count', 0)
                }

            return LLMResponse(
                content=content,
                model=self._model_name,
                usage=usage,
                structured_data=structured_data,
                raw_response=response
            )

        except Exception as e:
            raise RuntimeError(f"Gemini API error: {str(e)}")

    def health_check(self) -> bool:
        """Check if Gemini API is accessible."""
        try:
            self._ensure_client()
            import google.generativeai as genai

            # Try to list models as a health check
            models = list(genai.list_models())
            return len(models) > 0
        except Exception:
            return False

    def get_model_info(self) -> Dict[str, Any]:
        """Get information about the Gemini model."""
        return {
            'backend': 'gemini',
            'model': self._model_name,
            'provider': 'Google',
            'supports_json_schema': True,
            'supports_vision': 'vision' in self._model_name.lower() or 'gemini-2' in self._model_name
        }